<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# для того, чтобы не выставлять параметры подключения в открытый доступ, 
# считаем параметры подкючения из тектового файла:
params = {k:v.strip() for k, v in (l.split(' = ') for l in open('config.txt'))} 

In [3]:
# подключение к базе:
connection = psycopg2.connect(
    dbname = params['dbname'],
    user = params['user'],
    password = params['password'],
    host = params['host'],
    port = params['port']
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [17]:
# текст запроса
query_3_1 = '''select count(id) as "количество вакансий"
                from vacancies
'''

In [18]:
# результат запроса
df_3_1 = pd.read_sql_query(query_3_1, connection)
df_3_1

C:\Users\Asus\AppData\Local\Temp\ipykernel_1456\4045908245.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_1 = pd.read_sql_query(query_3_1, connection)


количество вакансий
0                49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [19]:
# текст запроса
query_3_2 = '''select count(id) as "количество работодателей" 
                from employers
'''

In [20]:
# результат запроса
df_3_2 = pd.read_sql_query(query_3_2, connection)
df_3_2

C:\Users\Asus\AppData\Local\Temp\ipykernel_1456\533540996.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_2 = pd.read_sql_query(query_3_2, connection)


количество работодателей
0                     23501

3. Посчитайте с помощью запроса количество регионов (таблица areas).

In [21]:
# текст запроса
query_3_3 = '''select count(id) as "количество регионов"
                from areas
'''

In [22]:
# результат запроса
df_3_3 = pd.read_sql_query(query_3_3, connection)
df_3_3

C:\Users\Asus\AppData\Local\Temp\ipykernel_1456\1395976625.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_3 = pd.read_sql_query(query_3_3, connection)


количество регионов
0                 1362

4. Посчитайте с помощью запроса количество сфер деятельности в базе (таблица industries).

In [23]:
# текст запроса
query_3_4 = '''select count(id) as "количество сфер деятельности"
                from industries
'''

In [24]:
# результат запроса
df_3_4 = pd.read_sql_query(query_3_4, connection)
df_3_4

C:\Users\Asus\AppData\Local\Temp\ipykernel_1456\854197215.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_4 = pd.read_sql_query(query_3_4, connection)


количество сфер деятельности
0                           294

***

### Выводы по предварительному анализу данных
Предварительный анализ данных проведен, по его итогам можно сказать, что данных для анализа достаточно: в базе около 49 тысяч вакансий от 23,5 тысяч работодателей, охватывающих 1362 региона и около трехсот сфер деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [27]:
# текст запроса
query_4_1 = '''select a.name as area,
                       count(v.id) as cnt_vacancies
               from vacancies v
                    left join areas a on v.area_id = a.id
               group by a.name
               order by count(v.id) desc
                
'''

In [26]:
# результат запроса
df_4_1 = pd.read_sql_query(query_4_1, connection)
df_4_1

C:\Users\Asus\AppData\Local\Temp\ipykernel_1456\620650199.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_1 = pd.read_sql_query(query_4_1, connection)


area  cnt_vacancies
0             Москва           5333
1    Санкт-Петербург           2851
2              Минск           2112
3        Новосибирск           2006
4             Алматы           1892
..               ...            ...
764       Тарко-Сале              1
765    Новоаннинский              1
766            Бирск              1
767           Сасово              1
768             Анна              1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [28]:
# текст запроса
query_4_2 = '''select count(id) 
               from vacancies
               where (salary_from is not null) or
                      (salary_to is not null)
'''

In [29]:
# результат запроса
df_4_2 = pd.read_sql_query(query_4_2, connection)
df_4_2

C:\Users\Asus\AppData\Local\Temp\ipykernel_1456\3746829199.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_2 = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [31]:
# текст запроса
query_4_3 = '''select round(avg(salary_from)) mean_salary_from,
                      round(avg(salary_to)) mean_salary_to
               from vacancies
'''

In [32]:
# результат запроса
df_4_3 = pd.read_sql_query(query_4_3, connection)
df_4_3

C:\Users\Asus\AppData\Local\Temp\ipykernel_1456\3595259731.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_3 = pd.read_sql_query(query_4_3, connection)


mean_salary_from  mean_salary_to
0           71065.0        110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [39]:
# текст запроса
query_4_4 = '''select schedule,
                       employment,
                       count(id) cnt_vacancies
                from vacancies
                group by 1, 2
                order by count(id) desc
'''

In [35]:
# результат запроса
df_4_4 = pd.read_sql_query(query_4_4, connection)
df_4_4

C:\Users\Asus\AppData\Local\Temp\ipykernel_1456\2028972612.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_4 = pd.read_sql_query(query_4_4, connection)


schedule           employment  cnt_vacancies
0        Полный день     Полная занятость          35367
1   Удаленная работа     Полная занятость           7802
2      Гибкий график     Полная занятость           1593
3   Удаленная работа  Частичная занятость           1312
4     Сменный график     Полная занятость            940
5        Полный день           Стажировка            569
6     Вахтовый метод     Полная занятость            367
7        Полный день  Частичная занятость            347
8      Гибкий график  Частичная занятость            312
9        Полный день     Проектная работа            141
10  Удаленная работа     Проектная работа            133
11     Гибкий график           Стажировка            116
12    Сменный график  Частичная занятость            101
13  Удаленная работа           Стажировка             64
14     Гибкий график     Проектная работа             18
15    Сменный график           Стажировка             12
16    Вахтовый метод     Проектная работа              2
17    Сменный график     Проектная работа              1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [40]:
# текст запроса
query_4_5 = '''select experience,
                       count(id) cnt_vacancies
                from vacancies
                group by experience
                order by 2 asc
'''

In [38]:
# результат запроса
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5

C:\Users\Asus\AppData\Local\Temp\ipykernel_1456\210463167.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_5 = pd.read_sql_query(query_4_5, connection)


experience  cnt_vacancies
0         Более 6 лет           1337
1           Нет опыта           7197
2       От 3 до 6 лет          14511
3  От 1 года до 3 лет          26152

***

### Выводы по детальному анализу вакансий
По детальному анализу вакансий можно сказать следующее:
- больше всего вакансий предлагается в Москве (около 5 тысяч), Санкт-Петербурге, Минске и городах-миллионниках (примерно по 2 тысячи), что логично, в больших городах всегда больше предложений по работе;
- заработная плата(верхняя или нижняя границы) указаны в менее чем 50 % вакансий (24073/49197*100% = 49%). Не очень приятная практика для соискателей, но как есть, многие работодатели предпочитают не указывать размер оплаты в открытую;
- средние значения верхней и нижней границы заработных плат находятся около 110 и 71 тысячи соответственно;
- преобладающее количество вакансий (около 72%) подразумевают полную занятость и полный рабочий день (35367/49197*100% = 72%), что логично, чаще всего требуются сотрудники, которые могут работать полную рабочую неделю;
- чаще всего требуются соискатели с опытом работы от 1 года до 3 лет, то есть хоть небольшой опыт, но нужен. 



# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [41]:
# текст запроса
query_5_1 = '''(select e.name,
                        'first' as place
                 from vacancies v
                      left join employers e on v.employer_id = e.id
                 group by e.name
                 order by count(v.id) desc
                 limit 1)
                 union
                 (select e_2.name,
                         'fifth'
                 from vacancies v_2
                      left join employers e_2 on v_2.employer_id = e_2.id
                 group by e_2.name
                 order by count(v_2.id) desc
                 offset 4 
                 limit 1)
                 order by 2 desc
'''

In [42]:
# результат запроса
df_5_1 = pd.read_sql_query(query_5_1, connection)
df_5_1

C:\Users\Asus\AppData\Local\Temp\ipykernel_1456\1147638897.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_1 = pd.read_sql_query(query_5_1, connection)


name  place
0         Яндекс  first
1  Газпром нефть  fifth

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [6]:
# текст запроса
query_5_2 = '''select a.name area,
                      count(distinct e.id) employers,
                      count(distinct v.id) vacancies
                from vacancies v  
                     full join areas a  on v.area_id = a.id
                     left join employers e on a.id = e.area
                where v.id is null 
                group by a.id
                order by count(distinct e.id) desc
'''

In [7]:
df_5_2 = pd.read_sql_query(query_5_2, connection)
df_5_2

C:\Users\Asus\AppData\Local\Temp\ipykernel_15316\762398461.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_2 = pd.read_sql_query(query_5_2, connection)


area  employers  vacancies
0                Россия        410          0
1             Казахстан        207          0
2    Московская область         75          0
3    Краснодарский край         19          0
4    Ростовская область         18          0
..                  ...        ...        ...
588             Скидель          0          0
589          Белоозерск          0          0
590           Орловский          0          0
591          Павловская          0          0
592             Таксимо          0          0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [8]:
# текст запроса
query_5_3 = '''select e.name,
                      count(distinct a.id) areas
               from employers e 
                    left join vacancies v on v.employer_id = e.id
                    left join areas a on v.area_id = a.id
                group by e.id
                order by 2 desc
'''

In [9]:
# результат запроса
df_5_3 = pd.read_sql_query(query_5_3, connection)
df_5_3

C:\Users\Asus\AppData\Local\Temp\ipykernel_15316\133400711.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_3 = pd.read_sql_query(query_5_3, connection)


name  areas
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
...                       ...    ...
23496              Но Смокинг      0
23497        МАСТЕРСКАЯ CREED      0
23498           Dzing Finance      0
23499                    КОРС      0
23500             Наука-Связь      0

[23501 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [10]:
# текст запроса
query_5_4 = '''select count(distinct e.id)
               from employers e 
                    left join employers_industries ei on ei.employer_id = e.id
               where industry_id is null
'''

In [11]:
# результат запроса
df_5_4 = pd.read_sql_query(query_5_4, connection)
df_5_4

C:\Users\Asus\AppData\Local\Temp\ipykernel_15316\1045119256.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_4 = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [13]:
# текст запроса
query_5_5 = '''select e.name,
                       count(ei.industry_id)
                from employers e 
                     left join employers_industries ei on ei.employer_id = e.id
                group by e.id
                      having count(ei.industry_id) = 4
                order by e.name
                offset 2 limit 1
'''

In [14]:
# результат запроса
df_5_5 = pd.read_sql_query(query_5_5, connection)
df_5_5

C:\Users\Asus\AppData\Local\Temp\ipykernel_15316\4025176131.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_5 = pd.read_sql_query(query_5_5, connection)


name  count
0  2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [15]:
# текст запроса
query_5_6 = f'''select count(e.id)
                from employers e 
                     left join employers_industries ei on ei.employer_id = e.id
                     left join industries i on ei.industry_id = i.id
                where i.name = 'Разработка программного обеспечения'
'''

In [16]:
# результат запроса
df_5_6 = pd.read_sql_query(query_5_6, connection)
df_5_6


C:\Users\Asus\AppData\Local\Temp\ipykernel_15316\1333455304.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_6 = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [17]:
# код для получения списка городов-милионников
# импорт необходимых библиотек
import requests
from bs4 import BeautifulSoup
# ссылку на список городов-миллионников копируем из браузера:
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
# получаем датафрейм с необходимой таблицей:
table_cities = pd.read_html(url)[1]
# преобразуем столбец 'Город' в таблице в кортеж:
cities = tuple(table_cities['Город'])


In [18]:
# текст запроса

query_5_7 = f'''select a.name as area,
                       count(v.id) as count
                from  vacancies v 
                left join employers e on e.id = v.employer_id
                left join areas a on v.area_id = a.id
                where e.name = 'Яндекс' and a.name in {cities}
                group by a.id
                union 
                select 'Total',
                        count(v_1.id)
                from  vacancies v_1 
                left join employers e_1 on e_1.id = v_1.employer_id
                left join areas a_1 on v_1.area_id = a_1.id
                where e_1.name = 'Яндекс' and a_1.name in {cities}
                order by count
'''

In [19]:
# результат запроса
df_5_7 = pd.read_sql_query(query_5_7, connection)
df_5_7

C:\Users\Asus\AppData\Local\Temp\ipykernel_15316\2595257991.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_7 = pd.read_sql_query(query_5_7, connection)


area  count
0              Омск     21
1         Челябинск     22
2        Красноярск     23
3         Волгоград     24
4             Пермь     25
5            Казань     25
6    Ростов-на-Дону     25
7               Уфа     26
8            Самара     26
9         Краснодар     30
10          Воронеж     32
11      Новосибирск     35
12  Нижний Новгород     36
13     Екатеринбург     39
14  Санкт-Петербург     42
15           Москва     54
16            Total    485

***

### Выводы по анализу работодателей
Проанализировав работодателей можно сказать следующее:
- у крупных компаний, таких как Яндекс и Газпромнефть количество вакансий больше, чем у менее крупных;
- таблица с регионами устроена так, что страна выделяется как отдельный регион. Видимо, когда работодатель не указывает конкретный город или регион, он указывает просто страну. Это надо учитывать при дальнейшей работе с датасетом. Так же видим, что в данных есть работодатели из разных стран;
- работодатели могут публиковать вакансии в нескольких регионах;
- некоторые работодатели не указывают сферу деятельности (примерно 35% от всего количества работодателей);
- у работодателей может быть указано несколько сфер деятельности;
- около 15 % работодателей указали в сфере деятельности "Разработка программного обеспечения"

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [20]:
# текст запроса
query_6_1 = '''select count(id)
                from vacancies
                where name like '%data%' or 
                      name like '%DATA%' or 
                    name like '%Data%' or 
                    name like '%данн%'  or 
                    name like '%Данн%'   or 
                    name like '%ДАНН%'
''' 

In [21]:
# результат запроса
df_6_1 = pd.read_sql_query(query_6_1, connection)
df_6_1

C:\Users\Asus\AppData\Local\Temp\ipykernel_15316\1281472032.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_1 = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [23]:
# текст запроса
query_6_2 = '''select count(id)
                from vacancies
                where (name ilike '%data scientist%' or 
                       name ilike '%data science%' or 
                       name ilike '%исследователь данных%' or 
                       (name like '%ML%'  and name not like '%HTML%') or
                       name ilike '%machine learning%'   or 
                       name ilike '%машинн%обучен%')
                       and (name ilike '%junior%' or 
                           experience = 'Нет опыта' or 
                           employment = 'Стажировка')
'''

In [24]:
# результат запроса
df_6_2 = pd.read_sql_query(query_6_2, connection)
df_6_2

C:\Users\Asus\AppData\Local\Temp\ipykernel_15316\118162426.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_2 = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [26]:
# текст запроса
query_6_3 = '''select count(id)
                from vacancies 
                where (name ilike '%data scientist%' or 
                       name ilike '%data science%' or 
                       name ilike '%исследователь данных%' or 
                       (name like '%ML%'  and name not like '%HTML%') or
                       name ilike '%machine learning%'   or 
                       name ilike '%машинн%обучен%')
                      and (key_skills ~* 'SQL|postgres')
'''

In [27]:
# результат запроса
df_6_3 = pd.read_sql_query(query_6_3, connection)
df_6_3

C:\Users\Asus\AppData\Local\Temp\ipykernel_15316\3310962512.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_3 = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [28]:
# текст запроса
query_6_4 = '''select count(id)
                from vacancies 
                where (name ilike '%data scientist%' or 
                       name ilike '%data science%' or 
                       name ilike '%исследователь данных%' or 
                       (name like '%ML%'  and name not like '%HTML%') or
                       name ilike '%machine learning%'   or 
                       name ilike '%машинн%обучен%')
                      and (key_skills ilike '%python%')
'''

In [29]:
# результат запроса
df_6_4 = pd.read_sql_query(query_6_4, connection)
df_6_4

C:\Users\Asus\AppData\Local\Temp\ipykernel_15316\614304167.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_4 = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [35]:
# текст запроса
query_6_5 = '''select round(avg((length(key_skills) - length(replace(key_skills, CHR(9), '')) + 1)), 2)   
                from vacancies 
                where (name ilike '%data scientist%' or 
                       name ilike '%data science%' or 
                       name ilike '%исследователь данных%' or 
                       (name like '%ML%'  and name not like '%HTML%') or
                       name ilike '%machine learning%'   or 
                       name ilike '%машинн%обучен%')
                       and  key_skills is not null
                       
'''

In [32]:
# результат запроса
df_6_5 = pd.read_sql_query(query_6_5, connection)
df_6_5

C:\Users\Asus\AppData\Local\Temp\ipykernel_15316\4089566367.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_5 = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [34]:
# текст запроса
query_6_6 = '''select experience,
                       round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to)))
                from vacancies 
                where (name ilike '%data scientist%' or 
                       name ilike '%data science%' or 
                       name ilike '%исследователь данных%' or 
                       (name like '%ML%'  and name not like '%HTML%') or
                       name ilike '%machine learning%'   or 
                       name ilike '%машинн%обучен%')
                       and  (salary_from is not null or salary_to is not null)
                group by experience                     
'''

In [19]:
# результат запроса
df_6_6 = pd.read_sql_query(query_6_6, connection)
df_6_6

C:\Users\Asus\AppData\Local\Temp\ipykernel_11108\2297777544.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_6 = pd.read_sql_query(query_6_6, connection)


experience     round
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

***

### Выводы по предметному анализу:
Из всего списка вакансий всего 1771 относятся к данным. Однако, стоит учитывать, что не все они действительно относятся к Data Science.
51 вакансия подходит для начинающих Data scientist-ов. Часто требуются знания SQL и Python. Зарплаты у начинающих специалистов Data Science существенно ниже, чем у более опытных.

### Дополнительные исследования:
Мне стало интересно, в каких странаах предлагаются вакансии в нашем датасете. Чтобы это узнать, я нашла список стран мира на Википедии и нашла, какие из них есть в таблице с вакансиями.

In [ ]:
# код для получения списка стран
# ссылку на список стран мира копируем из браузера:
url = 'https://ru.wikipedia.org/wiki/%D0%A1%D0%BF%D0%B8%D1%81%D0%BE%D0%BA_%D0%B3%D0%BE%D1%81%D1%83%D0%B4%D0%B0%D1%80%D1%81%D1%82%D0%B2'
# получаем датафрейм с необходимой таблицей:
table_countries = pd.read_html(url)[0]

# преобразуем столбец 'Страна' в таблице в кортеж:
countries = tuple(table_countries['Страна'])
countries

('Австралия',
 'Австрия',
 'Азербайджан',
 'Албания',
 'Алжир',
 'Ангола',
 'Андорра',
 'Антигуа и Барбуда',
 'Аргентина',
 'Армения',
 'Афганистан',
 'Багамы',
 'Бангладеш',
 'Барбадос',
 'Бахрейн',
 'Белоруссия',
 'Белиз',
 'Бельгия',
 'Бенин',
 'Болгария',
 'Боливия',
 'Босния и Герцеговина',
 'Ботсвана',
 'Бразилия',
 'Бруней',
 'Буркина-Фасо',
 'Бурунди',
 'Бутан',
 'Вануату',
 'Великобритания',
 'Венгрия',
 'Венесуэла',
 'Восточный Тимор',
 'Вьетнам',
 'Габон',
 'Гаити',
 'Гайана',
 'Гамбия',
 'Гана',
 'Гватемала',
 'Гвинея',
 'Гвинея-Бисау',
 'Германия',
 'Гондурас',
 'Гренада',
 'Греция',
 'Грузия',
 'Дания',
 'Джибути',
 'Доминика',
 'Доминикана',
 'Египет',
 'Замбия',
 'Зимбабве',
 'Израиль',
 'Индия',
 'Индонезия',
 'Иордания',
 'Ирак',
 'Иран',
 'Ирландия',
 'Исландия',
 'Испания',
 'Италия',
 'Йемен',
 'Кабо-Верде',
 'Казахстан',
 'Камбоджа',
 'Камерун',
 'Канада',
 'Катар',
 'Кения',
 'Кипр',
 'Киргизия',
 'Кирибати',
 'Китай',
 'Колумбия',
 'Коморы',
 'Конго',
 'ДР Конго

In [ ]:

query_add_1 = f'''select a.name as area,
                       count(v.id) as count
                from  vacancies v 
                left join areas a on v.area_id = a.id
                where a.name in {countries}
                group by a.name
                union 
                select 'Total',
                        count(v_1.id)
                from  vacancies v_1 
                left join areas a_1 on v_1.area_id = a_1.id
                where a_1.name in {countries}
                order by count desc
'''

In [ ]:
# результат запроса
df_add_1 = pd.read_sql_query(query_add_1, connection)
df_add_1

C:\Users\Asus\AppData\Local\Temp\ipykernel_1456\340634343.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_add_1 = pd.read_sql_query(query_add_1, connection)


area  count
0            Total    632
1          Армения    223
2             Кипр    136
3           Польша     53
4           Турция     39
5           Сербия     26
6         Германия     19
7            Литва     19
8       Черногория     14
9            Чехия     14
10             ОАЭ     12
11             США     12
12        Болгария     12
13          Латвия     10
14         Венгрия      9
15       Финляндия      7
16         Испания      4
17     Таджикистан      4
18         Австрия      4
19         Эстония      3
20      Нидерланды      2
21           Индия      2
22  Великобритания      2
23         Израиль      2
24           Китай      1
25        Словакия      1
26       Индонезия      1
27         Румыния      1

# Общий вывод по проекту

- Данных в проекте, в целом много.
- Некоторые данные желательно упорядочить. В частности то, что в регионах встречаются как страны, так и сами регионы стран. То есть, работодатели указывают не конкретный регион, например, Москва, а просто страну, например, Россия.
- Вакансий, относящихся к Data Science немного относительно всего датасета. Проверить их действительную расположенность к Data Science сложно автоматическию

In [4]:
# не забываем закрыть соединение после окончания работы
connection.close()